### PR TITLE
Canonicalize positional arguments

### PR DESCRIPTION
**Pull request by peterbell10**
As suggested in #167

### CODE REVIEW

File: uarray/backend.py [hunk 1]
```diff
@@ -174,12 +174,25 @@ def replace_dispatchables(
...
         kwargs = {k: kwargs[k] for k in opts if k in kwargs}
-        for k, v in defaults.items():
+        for k, v in kw_defaults.items():
             if k in kwargs and kwargs[k] is v:
                 del kwargs[k]
 
         return args, kwargs
 
+    def canonicalize_args(args):
+        if len(args) > len(arg_defaults):
```
Comment (hameerabbasi) [inline uarray/backend.py]: Isn't this branch already handled by the `return` at the end of this function?

File: uarray/backend.py [hunk 2]
```diff
@@ -174,12 +174,25 @@ def replace_dispatchables(
         args, kwargs = argument_replacer(args, kwargs, tuple(replaced_args))
 
         kwargs = {k: kwargs[k] for k in opts if k in kwargs}
```
Comment (hameerabbasi) [inline uarray/backend.py]: It might be nice to factor this out into the same function as `*args`.

File: uarray/backend.py [hunk 3]
```diff
@@ -309,14 +322,20 @@ def skip_backend(backend):
...
     sig = inspect.signature(f)
-    defaults = {}
+    kw_defaults = {}
+    arg_defaults = []
     opts = set()
     for k, v in sig.parameters.items():
         if v.default is not inspect.Parameter.empty:
-            defaults[k] = v.default
+            kw_defaults[k] = v.default
+        if v.kind in (
```
Comment (hameerabbasi) [inline uarray/backend.py]: Nice check. ;) I wasn't sure you'd add it on the first pass.

File: uarray/backend.py (hunk 1 above)
Comment (peterbell10) [reply uarray/backend.py]: The issue is with this line
```python
for a, d in zip(args[::-1], arg_defaults[len(args) - 1 :: -1]):
```
the slice of `arg_defaults` will not match up with `args`.

File: uarray/backend.py (hunk 2 above)
Comment (peterbell10) [reply uarray/backend.py]: This step is done after argument replacement whereas I have put my step before `args` reaches the backend. This way it's only done once.

**Comment by hameerabbasi** [comment]
Are you going to make any more changes? If not I'll merge. 😄

**Comment by peterbell10** [comment]
I think it's good

**Comment by hameerabbasi** [comment]
Thanks for this, @peterbell10!

**Pull request merged**